### DISCUSSION TITLE
How to manage part assemblies with parts purchased by assembly house and some as consigned parts

### DISCUSSION THREAD

**Discussion by dspmathguru** [inventree/InvenTree#6395] [Q&A]
We typically try to purchase and negotiate our PCBs and the expensive parts in on our boards with the manufacturer of the part.   At the same time we like to manage and to be able to see the total cost of all parts in a board build.  Thus we have put the DigiKey/Mouser pricing for all parts in our database and in the BOM for a board.   So we have a 'high' cost or a 'qty 1' cost for a board build.

When we build a board we only consign some of the parts to the assembly house.  Which means most parts are being purchased for the build by the assembly house and are part of the assembly house costs, but then we consign a few expensive parts (like PCBs, FPGAs, .etc.) to them.  

We would like to keep track of what we consign and to also know the price of a board build.  From build to build we consign different parts based on our negotiations with the assembly house, parts manufactures, distributors, etc.   

Does anyone have advice on how to manage getting the cost of a build where some parts are built into the assembly price and other parts that we consign are actually part of the COGS.

We do not want to double count parts we do not consign in the COGS for a board (assembly house + in the BOM of the board).

Is our only choice to have a modified BOM that only has the parts we consign in them?  Or some other idea ????  Thank you in advance.

**Comment by SchrodingersGat** [comment]
@dspmathguru if I am reading your question correctly, I think there are two parts:

**A) How to track the cost of a build**

The true cost of stock allocated to a build order is not currently calculated - rather an overall "BOM cost" is estimated based on available pricing data. It would not be overly difficult to calculate the true "assembly cost" of a build order, as long as the allocated stock has correct pricing information.

- Related issue: https://github.com/inventree/InvenTree/issues/5481

**B) How to handle an external build order**

This has been discussed before but without a good resolution. Further discussion here may lead to a good idea for how to implement this "cleanly" - but it has been on the "to do" pile for a while now:

- Related issue: https://github.com/inventree/InvenTree/issues/1071

**Reply by dspmathguru** [reply]
B - has elements of exactly what we need.   I would like to figure out how to work on this.

I've struggled to figure out how to start being a developer on this project.  I'm a very experienced Python coder (data science, medical imaging products, hardware interfaces with python, ...) and computer scientist.   I have no experience with Django or web development.  I'm lost in the structure of the code base and tool flow.  I easily created external python tools to add data to the inventree database from our external part databases we use for creating PCBs using the python API.   I just need to get my arms around how to add features that create actions within your development structure.   Any pointers would be great.

Here are some features I see we need, please comment and suggest where these go in the issue database and if they even make sense:
- Stock locations need addresses and a contact that manages them.  Then can be used to create PO's with appropriate location to ship product and other uses.
- Add a tag to a specific build's BOM item that says if the part is consigned.   Default to not consigned on tags that are part of the electronic part category tree.  Otherwise default to consigned.
- Add to the build web page a way to check consigned, not-consigned easily.
- Add a way on the Build page to select supplier for BOM items that have multiple and consigned items
- Generate PO's per supplier
- Go to each PO and update with additional quantities and items as seen fit
- Issue the PO's
- Receive finished goods

**Reply by matmair** [reply]
At least half the suggested features here are already present, maybe read the docs [regarding orders](https://docs.inventree.org/en/stable/order/purchase_order/) the PO stuff is covered.

**Reply by dspmathguru** [reply]
Correct, but a solution needs to be complete not half complete.

**Reply by matmair** [reply]
Creating issues for existing features is not how I develop software but maybe we develop software differently. I would prefer not to raise missing features before the docs were consulted.

**Reply by dspmathguru** [reply]
I see, I listed a process of steps, some are done, some are not.  I asked how I can get into writing the code myself.  Cherry-picking a set of issues out of a design solution comes second after understand the solution.  I do work out the steps first and then second create a set of narrowly defined issues that address the parts that need to be created.  I did not put this discussion in the issue area, but the discussion area as I figured we could discuss the best approach, then create a set of issues and then also give me some advice on how to start.

I'm not asking for the work to be done for me, I can do it.  Though I'd like to be sure I have the steps figured out with advice from the experts on this system.

I have read the doc's and we have 5K+ parts in the database and are creating Purchase Orders with a group of engineers and buyers outside of myself.  We are also creating build orders.  We have jumped in pretty quickly as this tool is solving a ton of problems for us and it is fitting naturally into our work flows.  Basically our engineering and operations are already using Inventree with great success.

Now we want to improve on our work flow and thus the discussion.

**Comment by SchrodingersGat** [comment]
@dspmathguru glad to hear that InvenTree is providing value to your company! I would suggest that the best approach here is to break the feature requests into attainable "bite sized" chunks where you can, and create a feature request issue for each one :)
